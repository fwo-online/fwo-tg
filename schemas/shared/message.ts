import type { CharacterPublic } from '@/character/characterPublic';
import type { Character } from '@/character';
import type { Player } from './player';
import type { GameStatus } from '@/game';
import type { Action } from './action';
import type { RPC } from './rpc';

export type ClientToServerMessage = Message<{
  character: [callback: (character: Character) => void];
  'lobby:enter': [callback: (characters: CharacterPublic[]) => void];
  'lobby:leave': [];
  'lobby:start': [];
  'lobby:stop': [];
  'game:connected': [callback: (payload: RPC<{ players: Record<string, Player> }>) => void];
  'game:order': [
    order: {
      power: number;
      action: string;
      target: string;
    },
    callback: (
      payload: RPC<{
        actions: Action[];
        magics: Action[];
        skills: Action[];
        power: number;
        orders: {
          power: number;
          action: string;
          target: string;
        }[];
      }>,
    ) => void,
  ];
}>;

type Message<T extends Record<string, unknown[]>> = {
  [K in keyof T]: (...args: T[K]) => void;
};

export type ServerToClientMessage = Message<{
  character: [character: Character];
  'lobby:enter': [character: CharacterPublic];
  'lobby:leave': [character: CharacterPublic];
  'lobby:list': [characters: CharacterPublic[]];
  'lobby:start': [character: CharacterPublic];
  'lobby:stop': [character: CharacterPublic];
  'game:start': [gameID: string];
  'game:end': [];
  'game:startOrders': [actions: { actions: Action[]; magics: Action[]; skills: Action[] }];
  'game:endOrders': [];
  'game:startRound': [
    {
      round: number;
      status: Partial<Record<string, GameStatus[]>>;
    },
  ];
  'game:endRound': [{ dead: Player[] }];
  'game:kick': [{ reason: string; player: Player }];
  'game:preKick': [{ reason: string; player: Player }];
}>;
