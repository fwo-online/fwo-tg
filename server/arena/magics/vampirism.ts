import { DmgMagic } from '../Constuructors/DmgMagicConstructor';

/**
 * Вампиризм
 * Основное описание магии общее требовани есть в конструкторе
 */
class Vampirism extends DmgMagic {
  constructor() {
    super({
      name: 'vampirism',
      displayName: 'Вампиризм',
      desc: 'Возвращает часть нанесеного урона в качесте жизней',
      cost: 12,
      baseExp: 10,
      costType: 'mp',
      lvl: 3,
      orderType: 'enemy',
      aoeType: 'target',
      magType: 'bad',
      chance: [92, 94, 95],
      effect: ['1d6+4', '1d5+5', '1d4+6'],
      dmgType: 'clear',
      profList: ['m'],
    });
  }

  /**
   * Основная функция запуска магии
   */
  run(): void {
    const { target, initiator } = this.params;
    target.stats.mode('down', 'hp', this.effectVal());
    initiator.stats.mode('up', 'hp', this.status.effect);
  }
}

export default new Vampirism();
