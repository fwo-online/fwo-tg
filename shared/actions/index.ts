import type { CharacterClass } from '@/character';

export type Magic = {
  name: string;
  description: string;
  displayName: string;
  effect: string[];
  cost: number;
  costType: CostType;
  lvl: number;
  orderType: OrderType;
  effectType?: EffectType;
};

export type Skill = {
  name: string;
  displayName: string;
  description: string;
  classList: Partial<Record<CharacterClass, number>>;
  bonusCost: number[];
  orderType: OrderType;
  cost: number;
  costType: CostType;
};

export enum CostType {
  MP = 'mp',
  EN = 'en',
}

export enum OrderType {
  All = 'all',
  Any = 'any',
  Enemy = 'enemy',
  Self = 'self',
  Team = 'team',
  TeamExceptSelf = 'teamExceptSelf',
}

export enum EffectType {
  Acid = 'acid',
  Fire = 'fire',
  Lighting = 'lighting',
  Frost = 'frost',
  Physical = 'physical',
  Clear = 'clear',
}

export type ActionType =
  | 'magic'
  | 'dmg-magic'
  | 'dmg-magic-long'
  | 'aoe-dmg-magic'
  | 'magic-long'
  | 'skill'
  | 'phys'
  | 'heal-magic'
  | 'heal'
  | 'protect'
  | 'dodge'
  | 'passive'
  | 'regeneration';

export type ReservedFailReason =
  | 'NO_INITIATOR'
  | 'NO_TARGET'
  | 'NO_MANA'
  | 'NO_ENERGY'
  | 'CHANCE_FAIL'
  | 'GOD_FAIL'
  | 'HEAL_FAIL'
  | 'SKILL_FAIL'
  | 'PHYS_FAIL'
  | 'NO_WEAPON';
