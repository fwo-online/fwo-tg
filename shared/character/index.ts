import type { CharacterAttributes } from './characterAttributesSchema';
import { CharacterClass } from './characterClassSchema';
import type { Attributes } from '@/shared/attributes';

export const characterAttributesByClass: Record<CharacterClass, CharacterAttributes> = {
  [CharacterClass.Warrior]: {
    str: 10,
    dex: 8,
    int: 3,
    wis: 3,
    con: 6,
  },
  [CharacterClass.Archer]: {
    str: 3,
    dex: 10,
    int: 8,
    wis: 3,
    con: 6,
  },
  [CharacterClass.Mage]: {
    str: 3,
    dex: 3,
    int: 8,
    wis: 10,
    con: 6,
  },
  [CharacterClass.Priest]: {
    str: 3,
    dex: 3,
    int: 10,
    wis: 8,
    con: 6,
  },
};

export type CharacterDynamicAttributes = Attributes & {
  maxTarget: number;
  spellLength: number;
};

export * from './characterAttributesSchema';
export * from './characterClassSchema';
export * from './characterSchema';
export * from './createCharacterSchema';
export * from './characterPublic';
