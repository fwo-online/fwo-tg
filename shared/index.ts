export * from './character';
export * from './inventory';
export * from './item';
export * from './itemSet';
export * from './shared';
export * from './actions';
export * from './game';
export * from './clan';
