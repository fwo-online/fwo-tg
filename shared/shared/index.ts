export * from './id';
export * from './modifiers';
export * from './player';
export * from './orderSchema';
export * from './message';
export * from './attributes';
export * from './action';
