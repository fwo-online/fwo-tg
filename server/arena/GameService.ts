import { createGame } from '@/api/game';
import type { Game } from '@/models/game';
import type { LongItem } from '@/arena/Constuructors/LongMagicConstructor';
import { engine } from '@/arena/EngineService';
import { HistoryService, type HistoryItem } from '@/arena/HistoryService';
import type * as magics from '@/arena/magics';
import OrderService from '@/arena/OrderService';
import PlayersService, { type Player } from '@/arena/PlayersService';
import { RoundService, RoundStatus } from '@/arena/RoundService';
import arena from '@/arena';
import { mapValues } from 'es-toolkit';
import EventEmitter from 'node:events';
import { type GameStatus, reservedClanName } from '@fwo/shared';
import { getRandomComponent } from '@/utils/getRandomComponent';

export type KickReason = 'afk' | 'run';

export interface GlobalFlags {
  isEclipsed: { initiator: Player }[];
}

/**
 * GameService
 *
 * @description Обработка около игровой логики
 * @module Service/Game
 * @todo сейчас после того как Player отключился, socket выходит из room.
 * Нужен механизм подключения обратно, если клиент "обновил" страницу или
 * переподключился к игре после disconnect(разрыв соединения)
 */

/**
 * Класс для объекта игры
 */
export default class GameService extends EventEmitter<{
  end: [
    {
      reason: string | undefined;
      statistic: Partial<Record<string, { exp: number; gold: number; nick: string }[]>>;
    },
  ];
  startOrders: [];
  endOrders: [];
  preKick: [{ reason: string; player: Player }];
  kick: [{ reason: string; player: Player }];
  startRound: [{ round: number; status: Record<string, GameStatus[]> }];
  endRound: [{ dead: Player[]; log: HistoryItem[] }];
}> {
  players: PlayersService;
  orders: OrderService;
  round = new RoundService();
  history = new HistoryService();
  longActions: Partial<Record<keyof typeof magics, LongItem[]>> = {};
  info!: Game;
  flags: {
    noDamageRound: number;
    global: GlobalFlags;
  };

  /**
   * Конструктор объекта игры
   * @param players массив игроков
   */
  constructor(players: string[]) {
    super();

    this.players = new PlayersService(players);
    this.orders = new OrderService(this.players, this.round);
    this.flags = {
      noDamageRound: 0,
      global: {
        isEclipsed: [],
      },
    };
  }

  /**
   * Функция проверки окончания игры
   */
  get isGameEnd(): boolean {
    return (
      this.isTeamWin ||
      this.players.alivePlayers.length === 0 ||
      this.flags.noDamageRound > 2 ||
      this.round.count > 9
    );
  }

  get isTeamWin(): boolean {
    const { [reservedClanName]: noClan, ...groupByClan } = this.players.groupByClan(
      this.players.alivePlayers,
    );
    if (!noClan?.length) {
      return Object.keys(groupByClan).length === 1;
    }
    return noClan.length === 1 && !Object.keys(groupByClan).length;
  }

  getEndGameReason() {
    if (this.flags.noDamageRound > 2) {
      return 'noDamageRound';
    }
  }

  get checkRoundDamage(): boolean {
    return !!this.history.hasDamageForRound(this.round.count);
  }

  /**
   * Предзагрузка игры
   */
  preLoading(): void {
    this.initHandlers();
    this.startGame();

    arena.games[this.info.id] = this;

    this.info.players.forEach((playerId) => {
      arena.characters[playerId].gameId = this.info.id;
    });
    // @todo add statistic +1 game for all players
  }

  /**
   * Старт игры
   */
  startGame(): void {
    console.debug('GC debug:: startGame', 'gameId:', this.info.id);
    this.round.initRound();
  }

  /**
   *@todo Остановка игры
   */
  pauseGame(): void {
    console.debug(this.info.id);
  }

  /**
   * Прекик, помечаем что пользователь не выполнил заказ и дальше будет выброшен
   * @param id id игрока, который будет помочен как бездействующий
   * @param reason строка, подставляющаяся в флаг isKicked
   */
  preKick(id: string, reason: KickReason): void {
    const player = this.players.getById(id);
    if (!player) {
      console.log('GC debug:: preKick', id, 'no player');
      return;
    }
    player.preKick(reason);
    this.emit('preKick', { reason, player });
  }

  /**
   * Функция "выброса игрока" из игры без сохранения накопленных статов
   * @param id id игрока, который будет выброшен
   * @param reason причина кика
   */
  kick(id: string, reason: KickReason): void {
    const player = this.players.getById(id);
    if (!player) {
      console.log('GC debug:: kick', id, 'no player');
      return;
    }

    this.emit('kick', { reason, player });

    const char = arena.characters[id];
    char.addGameStat({ runs: 1 });
    void char.saveToDb();
    char.autoreg = false;
    this.players.kick(id);
    this.info.players.splice(this.info.players.indexOf(id), 1);
    this.cleanLongMagics();
    char.gameId = '';
  }

  checkRun(player: Player) {
    if (player.flags.isKicked === 'run') {
      this.kick(player.id, player.flags.isKicked);
      return;
    }
  }

  /**
   * Проверяем делал ли игрок заказ. Помечает isKicked, если нет
   * @param player
   */
  checkOrders(player: Player): void {
    if (!player.alive || this.round.count === 1) {
      return;
    }

    if (!this.orders.checkPlayerOrderLastRound(player.id)) {
      if (player.flags.isKicked === 'afk') {
        this.kick(player.id, 'afk');
      } else {
        this.preKick(player.id, 'afk');
      }
    } else {
      /** @todo create clear kick flag method */
      player.preKick(undefined);
    }
  }

  /**
   * Ставим флаги, влияющие на окончание игры
   */
  handleEndGameFlags(): void {
    if (this.checkRoundDamage) {
      this.flags.noDamageRound = 0;
    } else {
      this.flags.noDamageRound += 1;
    }
  }

  recordOrderResult(item: HistoryItem) {
    this.history.addHistoryForRound(item, this.round.count);
  }

  getLastRoundResults() {
    return this.history.getHistoryForRound(this.round.count - 1);
  }

  getRoundResults() {
    return this.history.getHistoryForRound(this.round.count);
  }

  /**
   * @description Завершение игры
   *
   */
  endGame(): void {
    console.log('GC debug:: endGame', this.info.id);
    // Отправляем статистику
    setTimeout(async () => {
      const statistic = this.statistic();
      await this.saveGame();
      this.forAllPlayers((player: Player) => {
        arena.characters[player.id].gameId = '';
      });

      this.emit('end', { reason: this.getEndGameReason(), statistic });
      this.removeAllListeners();

      arena.mm.cancel();

      // FIXME move to client side
      this.forAllPlayers((player: Player) => {
        const char = arena.characters[player.id];
        if (char.expEarnedToday >= char.expLimitToday) {
          char.autoreg = false;
        }
        if (!char.autoreg) return;
        arena.mm.push({
          id: player.id,
          psr: 1000,
          startTime: Date.now(),
        });
      });
    }, 5000);
  }

  /**
   * Создание объекта в базе // потребуется для ведения истории
   * @return Объект созданный в базе
   */
  async createGame() {
    try {
      const dbGame = await createGame(this.players.init);
      this.info = dbGame;
      this.info.id = this.info._id.toString();
      return this;
    } catch (e) {
      console.log('GC debug:: createGame', e);
    } finally {
      this.preLoading();
    }
  }

  /**
   * Очищаем глобальные флаги в бою
   * затмение, бунт богов, и т.п
   */
  refreshRoundFlags(): void {
    this.flags.global.isEclipsed = [];
  }

  /**
   * Подвес
   */
  initHandlers(): void {
    // Обработка сообщений от Round Module
    this.round.subscribe(({ state, round }) => {
      switch (state) {
        case RoundStatus.START_ROUND: {
          this.forAllPlayers(this.checkOrders);
          this.sendStatus(round);
          break;
        }
        case RoundStatus.END_ROUND: {
          this.forAllPlayers(this.checkRun);
          this.emit('endRound', { dead: this.sortDead(), log: this.getRoundResults() });
          this.players.reset();
          this.orders.reset();
          this.handleEndGameFlags();
          if (this.isGameEnd) {
            this.round.unsubscribe();
            this.endGame();
          } else {
            this.refreshRoundFlags();
            this.round.nextRound();
          }
          break;
        }
        case RoundStatus.ENGINE: {
          engine(this);
          break;
        }
        case RoundStatus.START_ORDERS: {
          this.emit('startOrders');
          break;
        }
        case RoundStatus.END_ORDERS: {
          this.emit('endOrders');
          break;
        }
        default: {
          console.log('InitHandler:', state, 'undef event');
        }
      }
    });
  }

  /**
   * Метод сохраняющий накопленную статистику игроков в базу и сharObj
   * @todo нужен общий метод сохраняющий всю статистику
   */
  async saveGame() {
    try {
      const promises = this.info.players.map(async (id) => {
        const player = this.players.getById(id);
        if (!player) {
          return;
        }
        const isWinner = player.alive;

        await arena.characters[id].resources.addResources({
          gold: player.stats.collect.gold,
          exp: player.stats.collect.exp,
          components: player.stats.collect.component
            ? { [player.stats.collect.component]: 1 }
            : undefined,
        });

        const kills = this.players.getKills(id).length;
        const death = isWinner ? 0 : 1;

        arena.characters[id].addGameStat({
          games: 1,
          death,
          kills,
        });
        return arena.characters[id].saveToDb();
      });

      await Promise.all(promises);
    } catch (e) {
      console.log('Game:', e);
    }
  }

  /**
   * Функция послематчевой статистики
   * @todo надо как-то переделать функцию, потому что непонятно, что она мутирует данные
   * @return возвращает строку статистики по всем игрокам
   */
  statistic() {
    this.giveGoldforKill();
    const winners = this.players.alivePlayers;
    const gold = this.players.deadPlayers.length ? 5 : 1;
    winners.forEach((p) => {
      p.stats.addGold(gold);
      p.stats.addComponent(getRandomComponent(60));
    });

    this.players.deadPlayers.forEach((p) => {
      p.stats.addComponent(getRandomComponent(20));
    });

    const playersByClan = this.players.groupByClan();

    const getStatisticString = (p: Player) => ({
      nick: p.nick,
      exp: p.stats.collect.exp,
      gold: p.stats.collect.gold,
      component: p.stats.collect.component,
    });

    const statisticByClan = mapValues(playersByClan, (players) => {
      return players?.map(getStatisticString);
    });

    return statisticByClan;
  }

  /**
   * Функция пробегает всех убитых и раздает золото убийцам
   */
  giveGoldforKill(): void {
    this.players.deadPlayers.forEach((p) => {
      const killer = this.players.getById(p.getKiller());
      if (killer && killer.id !== p.id) {
        killer.stats.addGold(5 * p.lvl);
      }
    });
  }

  /**
   * Функция выставляет "смерть" для игроков имеющих hp < 0;
   * Отсылает сообщение о смерти игрока в последнем раунде
   * @todo сообщение о смерти как-то нормально нужно сделать,
   * чтобы выводило от чего и от кого умер игрок
   */
  sortDead() {
    const dead = this.players.sortDead();
    this.cleanLongMagics();
    return dead;
  }

  /**
   * Очистка массива длительных магий от умерших
   */
  cleanLongMagics(): void {
    this.longActions = mapValues(this.longActions, (longMagicType) => {
      return longMagicType?.filter((act) => {
        return this.players.getById(act.target)?.alive;
      });
    });
  }

  /**
   * Интерфейс для работы со всеми игроками в игре
   * @param f функция применяющая ко всем игрокам в игре
   */
  forAllPlayers(f: (player: Player) => void): void {
    this.players.players.forEach((p) => f.call(this, p));
  }

  getStatus() {
    const playersByClan = this.players.groupByClan();

    return mapValues(playersByClan, (players = []) => {
      return players.map((p) => p.getStatus());
    });
  }

  /**
   * Рассылка состояний живым игрокам
   */
  sendStatus(round: number): void {
    const status = this.getStatus();

    this.emit('startRound', { round, status });
  }
}
