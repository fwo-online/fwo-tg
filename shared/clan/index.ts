import { nameSchema } from '@/shared';
import * as v from 'valibot';

export const reservedClanName = '__clan';

export type Clan = {
  id: string;
  name: string;
  gold: number;
  lvl: number;
  owner: string;
  players: string[];
  requests: string[];
  hasEmptySlot: boolean;
  maxPlayers: number;
};

export type ClanPublic = Pick<Clan, 'id' | 'name' | 'players' | 'owner'>;

export const createClanSchema = v.object({
  name: nameSchema,
});

export const clanLvlCost = [100, 250, 750, 1500];
export const clanAcceptCostPerLvl = 50;
