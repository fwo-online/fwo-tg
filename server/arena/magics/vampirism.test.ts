import {
  describe, beforeAll, beforeEach, afterEach, it, spyOn, expect,
} from 'bun:test';
import casual from 'casual';
import GameService from '@/arena/GameService';
import TestUtils from '@/utils/testUtils';
import vampirism from './vampirism';
import { CharacterClass } from '@fwo/shared';

describe('vampirism', () => {
  let game: GameService;

  beforeAll(() => {
    casual.seed(1);
  });

  beforeEach(async () => {
    const initiator = await TestUtils.createCharacter({ prof: CharacterClass.Mage, magics: { vampirism: 2 } });
    const target = await TestUtils.createCharacter();

    game = new GameService([initiator.id, target.id]);
  });

  beforeEach(() => {
    spyOn(global.Math, 'random').mockReturnValue(0.5);
  });

  afterEach(() => {
    spyOn(global.Math, 'random').mockRestore();
  });

  it('should take dmg and heal on same amount', () => {
    game.players.players[0].proc = 1;

    vampirism.cast(game.players.players[0], game.players.players[1], game);

    expect(game.players.players[0].stats.val('hp')).toBe(11.63)
    expect(game.players.players[1].stats.val('hp')).toBe(4.37)

    expect(TestUtils.normalizeRoundHistory(game.getRoundResults())).toMatchSnapshot();
  });

  it('should heal only above target`s 0 hp', () => {
    game.players.players[0].proc = 1;
    game.players.players[1].stats.set('hp', 1);

    vampirism.cast(game.players.players[0], game.players.players[1], game);

    expect(game.players.players[0].stats.val('hp')).toBe(9)
    expect(game.players.players[1].stats.val('hp')).toBe(-2.63)

    expect(TestUtils.normalizeRoundHistory(game.getRoundResults())).toMatchSnapshot();
  });

  it('should not heal whet target`s hp less than 0', () => {
    game.players.players[0].proc = 1;
    game.players.players[1].stats.set('hp', -1);

    vampirism.cast(game.players.players[0], game.players.players[1], game);

    expect(game.players.players[0].stats.val('hp')).toBe(8)
    expect(game.players.players[1].stats.val('hp')).toBe(-4.63)

    expect(TestUtils.normalizeRoundHistory(game.getRoundResults())).toMatchSnapshot();
  });
});
