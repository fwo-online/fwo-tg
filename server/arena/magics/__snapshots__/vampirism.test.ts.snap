// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`vampirism should take dmg and heal on same amount 1`] = `
"*–ò–≥—Ä–æ–∫ 1* —Å–æ—Ç–≤–æ—Ä–∏–ª _–í–∞–º–ø–∏—Ä–∏–∑–º_ –Ω–∞ *–ò–≥—Ä–æ–∫ 2* –Ω–∞–Ω–µ—Å—è *3.63* –∏ –≤–æ—Å—Å—Ç–∞–Ω–æ–≤–∏–≤ —Å–µ–±–µ *3.63* –∑–¥–æ—Ä–æ–≤—å—è
\\[ –ò–≥—Ä–æ–∫ 2 ‚ú® üíî-3.63/4.37 üìñ39 ]"
`;

exports[`vampirism should heal only above target\`s 0 hp 1`] = `
"*–ò–≥—Ä–æ–∫ 1* —Å–æ—Ç–≤–æ—Ä–∏–ª _–í–∞–º–ø–∏—Ä–∏–∑–º_ –Ω–∞ *–ò–≥—Ä–æ–∫ 2* –Ω–∞–Ω–µ—Å—è *3.63* –∏ –≤–æ—Å—Å—Ç–∞–Ω–æ–≤–∏–≤ —Å–µ–±–µ *1* –∑–¥–æ—Ä–æ–≤—å—è
\\[ –ò–≥—Ä–æ–∫ 2 ‚ú® üíî-3.63/-2.63 üìñ39 ]"
`;

exports[`vampirism should not heal whet target\`s hp less than 0 1`] = `
"*–ò–≥—Ä–æ–∫ 1* —Å–æ—Ç–≤–æ—Ä–∏–ª _–í–∞–º–ø–∏—Ä–∏–∑–º_ –Ω–∞ *–ò–≥—Ä–æ–∫ 2* –Ω–∞–Ω–µ—Å—è *3.63* –∏ –≤–æ—Å—Å—Ç–∞–Ω–æ–≤–∏–≤ —Å–µ–±–µ *0* –∑–¥–æ—Ä–æ–≤—å—è
\\[ –ò–≥—Ä–æ–∫ 2 ‚ú® üíî-3.63/-4.63 üìñ39 ]"
`;
